### DISCUSSION TITLE
Additional data field collection for 2025

### DISCUSSION THREAD

**Discussion by djong1** [djong1/gp3s#265] [GPS Technology (devices and formats)]
GPS Models, Chipsets and software versions varies and can have a positive or negative behaviour on the accuracy of speed data. Therefore we would like to collect more information from the track file itself next to the meta data collection that is already collected when uploading (manual or via API).

**Additional data fields to be collected**

### DEVICE INFO (for garmin only):

(to capture software version and GNSS version)
1. Manufacturer (Garmin)
2. product (Garmin)
3. software version (Garmin)

Conditions:
-  when device index == creator and index == 0 (Garmin, fr255)
- when device type == gps (GNSS, 10.00)

![image](https://github.com/user-attachments/assets/64c775dc-13c9-4e58-bcb6-707b401b16eb)

### File ID

1. manufacturer (Garmin, Coros)
2. product (Garmin, Coros)
3. serial number (Garmin only)

![image](https://github.com/user-attachments/assets/1358aeb0-9e2f-4749-ac92-68df36a6f591)

### Session

1. sport (Garmin, Coros)
2. sub sport (Garmin, Coros)

![image](https://github.com/user-attachments/assets/1db14b06-5d00-4041-95c6-f5e7852fed97)

### GPS Event
Garmin Fenix 7 and above only

1. mode

![image](https://github.com/user-attachments/assets/37c40b22-2181-4310-8cc7-c296a18530de)

**Comment by djong1** [comment]
@jhdebruin @Logiqx

**Comment by Logiqx** [comment]
I'll give each of the frame types as reference number so we can focus on them individually:

## 1. device_info - Garmin only

Worth noting that the device_info frame can appear multiple times in the FIT, and items can be duplicated. However, so long as the selection criteria are correct you'll get the right values.

I think device_type should be used for selection purposes; blank = watch, 0 = gps.

![image](https://github.com/user-attachments/assets/82f9762f-a4a8-45fc-b1e3-e55309ab0ece)

![image](https://github.com/user-attachments/assets/e7cc09de-25ca-4383-b2b9-1d608deea1ea)

The product (e.g. 3992 = fr255 or 3865 = gnss) and software version (e.g. 19.18) can then be recorded.

## 2. file_id

Garmin and COROS are consistent for manufacturer (string) and product (integer):
- manufacturer (e.g. garmin or coros) and product (e.g. 3992 = fr255, 822 = COROS APEX 2 Pro)
- serial_number (e.g. 3466382009) is unique to Garmin.

![image](https://github.com/user-attachments/assets/935c26f2-1912-4111-8bf5-7bf75ac56b4d)

![image](https://github.com/user-attachments/assets/bb2115e5-023a-4e93-a477-f88fdc80cf9d)

Worth noting that COROS fail to distinguish the APEX 2 and APEX 2 Pro - https://github.com/Logiqx/gp3s-coros/issues/36

## 3. session

- sport is available on COROS and Garmin but COROS uses windsurfing for speedsurfing + windsurfing
- sub_sport is only available on Garmin

![image](https://github.com/user-attachments/assets/0bd053ac-5e9c-4df5-b198-10658ddd3257)

![image](https://github.com/user-attachments/assets/54510900-6689-4af8-b704-8d96f300baa0)

I've suggested some additional items to Scott:
- app_name = APPro
- app_ver = v407-pre6
- craft = windsurf / windfoil / wingfoil / kitesurf / kitefoil

n.b. Craft can also be used for the sport name elsewhere in the FIT, so that Garmin connect shows sessions such as "Weymouth Wingfoiling".

![image](https://github.com/user-attachments/assets/7f09fc82-9750-465c-889e-71482c33204f)

## 4. gps_events - Garmin only

I've not seen these in Garmin FIT files. Can you provide an example?

## 5. developer_data_id - Garmin only

This relates to downloads from Garmin Connect. It is probably worth recording as it cannot be easily faked.

application_id (list of numbers), application_version (integer)

![image](https://github.com/user-attachments/assets/062ba9c6-152a-440f-b83a-c56bac9c5250)

## 6. field_description

Multiple field_description frames may be present in the FIT so they need to be combined (and maybe de-duplicated).

I think it is worth saving a list of all field_name values (but not units or data types).

![image](https://github.com/user-attachments/assets/577b5a0d-9871-4f5f-96fc-9224c9a9c42d)

![image](https://github.com/user-attachments/assets/273810ce-2c99-4c32-b004-00df612263ac)

Tip: The presence of developer fields can be used to distinguish between windsurfing and speedsurfing sessions on COROS.

**Comment by Logiqx** [comment]
I've written some notes in a better format for everyone working with Garmin + COROS files.

The various FIT records are described below and the insights that should (ideally) be retained.


## file_id

FIT files from COROS use the same product number for the APEX 2 and APEX 2 Pro. A simple workaround is proposed using the product_name field, mapping the APEX 2 to product number 821. More details about the issue are on [GitHub](https://github.com/Logiqx/gp3s-coros/issues/36).

| Name                       | Field / Value | Criteria                                            | Garmin? | COROS? |
| -------------------------- | ------------- | --------------------------------------------------- | :-----: | :----: |
| manufacturer <sup>1</sup>  | manufacturer  | -                                                   |    Y    |   Y    |
| product <sup>1 2</sup>     | product       | -                                                   |    Y    |   Y    |
|                            | 821           | product = 822 and product_name == "COROS Wearables" |    -    |   Y    |
| product_name <sup>1</sup>  | product_name  | -                                                   |    -    |   Y    |
|                            | "APEX 2"      | product = 822 and product_name == "COROS Wearables" |    -    |   Y    |
| serial_number <sup>3</sup> | serial_number | -                                                   |    Y    |   -    |

Notes:

1. Suunto and Timex also include manufacturer, product and product_name.
2. Garmin used to include garmin_product instead of product.
3. Timex also includes serial_number.



## developer_data_id

The application details from the developer_data_id record can be used to identify the application downloaded from the Connect IQ Store, but not side-loaded betas.

n.b. The application_id in COROS files doesn't provide any useful insights.

| Name                | Field               | Criteria | Garmin? | COROS? |
| ------------------- | ------------------- | -------- | :-----: | :----: |
| application_id      | application_id      | -        |    Y    |   Y    |
| application_version | application_version | -        |    Y    |   -    |



## session

The session record includes activity details that can be used to determine whether the Garmin settings meet the GP3S requirements.

| Name        | Field       | Criteria | Garmin? | COROS? |
| ----------- | ----------- | -------- | :-----: | :----: |
| sport       | sport       | -        |    Y    |   Y    |
| sub_sport   | sub_sport   | -        |    Y    |   -    |
| craft       | craft       | -        |  APPro  |   -    |
| app_name    | app_name    | -        |  APPro  |   -    |
| app_ver     | app_ver     | -        |  APPro  |   -    |
| gnss_config | gnss_config | -        |  APPro  |   -    |

Criteria for approved Garmin devices to receive the green icon on GP3S:

- sport is either 0 (SPORT_GENERIC), 43 (SPORT_WINDSURFING), or 44 (SPORT_KITESURFING)
- gnss_config is either 5 (CONFIGURATION_GPS_GLONASS_GALILEO_BEIDOU_L1) or 6 (CONFIGURATION_GPS_GLONASS_GALILEO_BEIDOU_L1_L5)

n.b. There is an additional approval process for individual watch models.



## field_description

The developer fields in a COROS file can be used to determine whether the session uses the speedsurfing mode.

| Name      | Value | Criteria (where exists)     | Garmin? |    COROS?    |
| --------- | ----- | --------------------------- | :-----: | :----------: |
| flag_hdop | True  | lower(field_name) == 'hdop' |    -    | speedsurfing |
| flag_cog  | True  | lower(field_name) == 'cog'  |    -    | speedsurfing |
| flag_sats | True  | lower(field_name) == 'sat'  |    -    | speedsurfing |
| flag_fix  | True  | field_name == 'fix_quality' |  APPro  |      -       |

Criteria for approved COROS devices to receive the green icon:

- One or more of flag_hdop, flag_cog, or flag_sats must be true
- This confirms whether the speedsurfing mode has been used, or not



## device_info

The firmware and GNSS version could prove to be extremely useful in the future, should any bad firmware be released.

n.b. Garmin watches back in early-mid 2022 experienced the issues that COROS have introduced to some of their Airoha-based watches during 2024.

| Name             | Field            | Criteria                                         | Garmin? | COROS? |
| ---------------- | ---------------- | ------------------------------------------------ | :-----: | :----: |
| firmware_version | software_version | device_type is empty, or device_index == creator |    Y    |   -    |
| gnss_version     | software_version | device_type == 0                                 |    Y    |   -    |

**Reply by djong1** [reply]
Thanks, we collect all fields now except the developer data. One addition we also collect "device" which is only available for Coros to collect the type of Coros watch.

**Reply by Logiqx** [reply]
No problem. I've also tweaked the notes to describe what Suunto do in their file_id records.

Like COROS, they also record the device name as text.

**Comment by Logiqx** [comment]
I’ve just discovered that https://www.fitfileviewer.com/ behaves completely differently in Chrome and Firefox. It’s a completely different app!

Now that I’ve been able to see GPS events in FIT files, I’ve determined what the different values / bits represent and documented it. These notes might be useful in the GP3S backend in the future, when trying to determine the GPS mode that was actually used for the session.

https://logiqx.github.io/gps-details/devices/garmin/developer/gps-events.html

Potentially this will allow people to use things like the GPSTC datafield, if that is their preference. Just need to be wary about SatIQ mode since the GPS mode will change automatically. Dual band to all systems would be fine, but not GPS only.
